이번 주에 다룰 내용은 아래와 같다.

1. 예외 처리
2. 문자열을 인쇄하는 다양한 방법 활용

#  예외 처리

프로그램 실행중에 오류(error)가 발생하면 예외(exception) 상태로 전환된다. 예외 상태로 전환되면 실행중이던 프로그램을 멈추고 발생한 오류를 알려준다. 즉, 예외는 오류를 처리하는 방식이며, C 언어에서는 존재하지 않는다.

예외 처리와 관련해서 다룰 내용은 다음과 같다.

* 예외의 종류 살펴보기
* `try ... except ...` 명령어를 이용하여 예외 처리하기
* `raise` 명령어를 이용하여 예외 발생시키기

### 예외 종류

가장 많이 발생하는 예외들을 살펴본다.

    ZeroDivisionError
    NameError
    IOError
    IndexError
    NotImplementedError
    
등등.

#### ZeroDivisionError

`0`으로 숫자를 나누려고 할 때 발생한다. 

In [1]:
4.5 / 0

ZeroDivisionError: float division by zero

#### NameError

선언되지 않은 변수를 사용하고자 할 때 발생한다. 

In [2]:
print(abc)

NameError: name 'abc' is not defined

#### IOError

아직 생성되지 않은 파일에 접근할 때, 저장 공간이 부족할 때, 저장이 제대로 되지 않을 때 등등 발생한다. 

In [3]:
f = open('not_yet_defined_file.txt', 'r') 

IOError: [Errno 2] No such file or directory: 'not_yet_defined_file.txt'

#### IndexError

시퀀스 자료형의 객체의 원소에 접근할 때 정해진 인덱스 범위를 초과할 때 발생한다.

In [4]:
l = range(3)
l[3]

IndexError: list index out of range

#### NotImplementedError

정의되지 함수를 사용할 때 고의로 에러를 발생키시고자 할 때 사용한다. `raise`를 이용한다. 

In [5]:
def my_complicated_function():
    """아주 복잡하지만 지금 당장 불필요"""
    raise NotImplementedError("아직 정의되어 있지 않음")

In [6]:
my_complicated_function()

NotImplementedError: 아직 정의되어 있지 않음

오류 처리를 사용하지 않으면 오류 메시지가 보이지 않을 수도 있음에 주의해야 한다.

In [7]:
def my_complicated_function_1():
    """아주 복잡하지만 지금 당장 불필요"""

In [8]:
my_complicated_function_1()

#### 오류의 종류

이외에 많은 종류의 오류가 있다.

* 보다 자세한 설명은 아래 사이트를 참조하면 된다. 
    
        https://docs.python.org/2/library/exceptions.html 
    
    
* 아니면 아래와 같이 `help` 명령어를 사용하여 확인할 수 있다. 

        >>> import exceptions
        >>> help(exceptions)
        
* 많은 종류의 오류를 무조건 외울 필요는 없다. 대신에 오류가 발생할 때마다 어떤 종류의 오류가 발생하는지 기억해두는 습관을 키워야 한다.

### `try ... except ...` 명령어 이용 예외 처리

오류가 발생하여 프로그램이 임의로 실행 중단되는 것을 방지하기 위해 사용한다. 

* 오류가 발생할 수 있는 명령어 부분을 `try` 문으로 감싼다.
* `try` 문으로 감싼 코드에서 오류가 발생할 경우 예외 처리하는 부분을 `except` 
    문으로 감싼다. 
* `try` 문으로 감싼 코드에서 오류가 발생하지 않으면 `except` 문으로 감싼 부분은 
    건너 뛴다. 

#### 기본 사용법

아래 코드를 실행하면 두 가지 경우가 발생할 수 있다.

* `myfilename.dat` 파일이 존재하는 경우 해당 파일을 열어 한 줄씩 보여준다.
* `myfilename.dat` 파일이 존재하지 않는 경우 아래와 같은 메시지가 보여지고 프로그램 실행이 멈춘다. 

        """해당 파일을 열수 없습니다. 프로그램을 더 이상 실행할 수 없습니다.
        An exception has occurred, use %tb to see the full traceback.
        SystemExit: 1"""
        
    - 위 메시지에서 한글 부분은 예외 처리 코드 중에서 두 개의 `print` 문이 실행된 결과이다.
    - `sys` 모듈의 `exit` 함수를 호출하면 프로그램이 중단된다. exit 함수가 숫자 `1`을 리턴하도록 하여 문제 때문에 프로그램이 중단되었음을 알리는 기능도 예외 처리에 포함시킨 것을 알 수 있다.
    - 디버깅을 위해 모두 필요한 정보들이 담기도록 예외 처리를 해야 한다. 

In [9]:
try:
    f = open('myfilename.dat', 'r') 
except:
    print("해당 파일을 열수 없습니다."), 
    print("프로그램을 더 이상 실행할 수 없습니다.")
    import sys
    sys.exit(1) #a way to exit the programle couldn’t be opened."), 

for line in f.readlines():
    print(line)

해당 파일을 열수 없습니다. 프로그램을 더 이상 실행할 수 없습니다.


SystemExit: 1

To exit: use 'exit', 'quit', or Ctrl-D.


#### `except` 문에 오류의 종류를 명시할 수 있다.

위 코드에서 `try`문에서 해당 파일이 존재하지 않으면 `IOError`가 발생하므로 `except IOError` 라고 명령해도 동일한 기능을 수행한다. 

In [10]:
try:
    f = open('myfilename.dat', 'r') 
except IOError:
    print("해당 파일을 열수 없습니다."), 
    print("프로그램을 더 이상 실행할 수 없습니다.")
    import sys
    sys.exit(1) #a way to exit the programle couldn’t be opened."), 

for line in f.readlines():
    print(line)

해당 파일을 열수 없습니다. 프로그램을 더 이상 실행할 수 없습니다.


SystemExit: 1

To exit: use 'exit', 'quit', or Ctrl-D.


#### `except` 문에 오류 종류와 오류메시지변수를 함께 사용해서 오류에 대한 보다 자세한 설명을 확인할 수 있다.

In [11]:
try:
    f = open('myfilename.dat', 'r') 
except IOError as er:
    print("해당 파일을 열수 없습니다."), 
    print("프로그램을 더 이상 실행할 수 없습니다.")
    print(er)
    import sys
    sys.exit(1) #a way to exit the programle couldn’t be opened."), 

for line in f.readlines():
    print(line)

해당 파일을 열수 없습니다. 프로그램을 더 이상 실행할 수 없습니다.
[Errno 2] No such file or directory: 'myfilename.dat'


SystemExit: 1

To exit: use 'exit', 'quit', or Ctrl-D.


#### `except` 문에 사용하는 오류 종류는 정확해야 한다.

아래 코드의 경우는 예외 처리가 제대로 되지 않는다. `try`문에서 발생하는 오류는 `ZeroDivisionError`인 반면에 `except IOError`는 `IOError`만을 처리하기 때문에 제대로 예외 처리를 못한다.

In [12]:
try:
    a = 1/0
except IOError:
    print("The file couldn’t be opened."), 
    print("This program stops here.")
    import sys
    sys.exit(1) #a way to exit the program


ZeroDivisionError: integer division or modulo by zero

#### `try ... except ....` 문의 보다 다양한 활용방법은 연습문제를 통해 습득할 예정이다.

* 아래 사이트 참조 가능
        https://wikidocs.net/30

## raise 함수

앞에서 `NotImplementedError`의 예제에서 보았듯이 `raise` 함수는 오류를 일부러 유발할 때 사용한다. 

In [13]:
raise NameError('HiThere')

NameError: HiThere

아래 코드에서처럼 `raise` 함수는 에러 인자를 선택적으로 받을 수 있다.

In [14]:
try:
    raise NameError('HiThere')
except NameError:
    print 'An exception flew by!'
    raise

An exception flew by!


NameError: HiThere

# 문자열을 인쇄하는 다양한 방법 활용

In [15]:
a = "string"
b = "string1"

In [16]:
print(a),; print(b)

string string1


위 경우 `a`와 `b`를 인쇄할 때 스페이스가 자동으로 추가된다. 
그런데 스페이스 없이 `stringstring1`으로 출력하려면 두 가지 방식이 있다.

* 문자열 덧셈 연산자를 활용한다.
* 서식이 있는 인쇄방식(formatted printing)을 사용한다. 

In [17]:
print(a+b)

stringstring1


In [18]:
print("{}{}".format(a,b))

stringstring1


서식이 있는 인쇄방식을 이용하여 다양한 형태로 자료들을 인쇄할 수 있다. 서식을 사용하는 방식은 크게 두 가지가 있다.

* `%` 를 이용하는 방식
    - C, Java 등에서 사용하는 방식과 거의 비슷하다.
    
* `format` 키워드를 사용하는 방식
    - 파이써 만의 새로운 방식이며 `%` 보다 좀 더 다양한 기능을 지원한다. 
    - Java의 경우 `MessageFormat` 클래스의 `format` 메소드가 비슷한 기능을 
    지원하지만 사용법이 좀 더 복잡하다. 

### 서식지정자(`%`)를  활용하여 문자열 인쇄하기

In [19]:
print("%s%s%d" % (a, b, 10))

stringstring110


In [20]:
from math import pi

In [21]:
print('%f' % pi)

3.141593


In [22]:
print("%.10f"% pi)

3.1415926536


In [23]:
print("%.30f"% pi)

3.141592653589793115997963468544


In [24]:
print("%.50f"% pi)

3.14159265358979311599796346854418516159057617187500


`pi`값 계산을 소숫점 이하 50자리 정도까지 계산함을 알 수 있다. 이것은 사용하는 컴퓨터의 한계이며 컴퓨터의 성능에 따라 계산능력이 달라진다. 

In [25]:
print("%f"% pi)
print("%f"% pi**3)
print("%f"% pi**10)

3.141593
31.006277
93648.047476


오른쪽으로 줄을 맞추려면 아래 방식을 사용한다.

* 숫자 12는 `pi**10`의 값인 `93648.047476`가 점(.)을 포함하여 총 12자리로 가장 길기에 선택되었다. 

* 표현방식이 달라지면 다른 값을 선택해야 한다.

In [26]:
print("%12f"% pi)
print("%12f"% pi**3)
print("%12f"% pi**10)

    3.141593
   31.006277
93648.047476


In [27]:
print("%16.10f"% pi)
print("%16.10f"% pi**3)
print("%16.10f"% pi**10)

    3.1415926536
   31.0062766803
93648.0474760830


비어 있는 자리를 숫자 `0`으로 채울 수도 있다.

In [28]:
print("%012f"% pi)
print("%012f"% pi**3)
print("%012f"% pi**10)

00003.141593
00031.006277
93648.047476


In [29]:
print("%016.10f"% pi)
print("%016.10f"% pi**3)
print("%016.10f"% pi**10)

00003.1415926536
00031.0062766803
93648.0474760830


자릿수는 계산결과를 예상하여 결정해야 한다. 
아래의 경우는 자릿수를 너무 작게 무시한다. 

In [30]:
print("%12.20f" % pi**19)

2791563949.59784364700317382812


### `format` 함수 사용하여 문자열 인쇄하기

`format` 함수를 이용하여 서식지정자(`%`)를 사용한 결과를 동일하게 구할 수 있다. 

In [31]:
print("{}{}{}".format(a, b, 10))

stringstring110


In [32]:
print("{:s}{:s}{:d}".format(a, b, 10))

stringstring110


In [33]:
print("{:f}".format(pi))
print("{:f}".format(pi**3))
print("{:f}".format(pi**10))

3.141593
31.006277
93648.047476


In [34]:
print("{:12f}".format(pi))
print("{:12f}".format(pi**3))
print("{:12f}".format(pi**10))

    3.141593
   31.006277
93648.047476


In [35]:
print("{:012f}".format(pi))
print("{:012f}".format(pi**3))
print("{:012f}".format(pi**10))

00003.141593
00031.006277
93648.047476


`format` 함수는 인덱싱 기능까지 지원한다. 

In [36]:
print("{2}{1}{0}".format(a, b, 10))

10string1string


인덱싱을 위해 키워드를 사용할 수도 있다.

In [37]:
print("{s1}{s2}{s1}".format(s1=a, s2=b, i1=10))

stringstring1string


In [38]:
print("{i1}{s2}{s1}".format(s1=a, s2=b, i1=10))

10string1string


인덱싱과 서식지정자를 함께 사용할 수 있다

In [39]:
print("{1:12f}, {0:12f}".format(pi, pi**3))

   31.006277,     3.141593


In [40]:
print("{p1:12f}, {p0:12f}".format(p0=pi, p1=pi**3))

   31.006277,     3.141593
